zerver/views/streams.py

```diff
@@ -315,12 +322,20 @@ def update_stream_backend(
         else:
             raise JsonableError(_("Invalid parameters"))
 
-    if is_private is not None:
-        # Default streams cannot be made private.
-        default_stream_ids = {s.id for s in get_default_streams_for_realm(stream.realm_id)}
-        if is_private and stream.id in default_stream_ids:
-            raise JsonableError(_("Default streams cannot be made private."))
+    # Ensure that a stream cannot be both a default stream for new users and private,
+    # and disallow making default streams private or private streams default.
+    if is_default_stream and is_private:
+        raise JsonableError(
+            _("A stream cannot be both a default stream for new users and private.")
+        )
 
+    if is_private and proposed_is_default_stream:
+        raise JsonableError(_("Default streams cannot be made private."))
+
+    if is_default_stream and proposed_is_private:
+        raise JsonableError(_("Private streams cannot be made default."))
```

Comment:
I think we can restructure this code in a better way. We can perhaps have a condtion like `if proposed_is_private and proposed_is_default_stream` as the outer if block condition and then condition for these 3 messages inside it. This will lead to grouping of the errors for the case that stream cannot be private and default at the same time. 

I am not sure what would be more readable, your thoughts on this?

Reply:
I think that's an improvement. I was thinking of writing them in if-else blocks to group these conditions, but this is better.